### ISSUE TITLE
Windows: Pen pressure detection works only once after start

### ISSUE THREAD

**Issue by Nick768** [xournalpp/xournalpp#3311]
**Affects versions :**
 - OS:
    Windows (7 and 11 tested)
 - Which version of libgtk do you use
    3.24.30
 - Version of Xournal++
    1.1.0 (commit: 9d1277d and fffaa72 [nightly] tested)
 - Installation method:
    Windows installer

**Describe the bug**
The pressure sensitivity detection of my Wacom Intuos S doesn't work after starting and drawing the first line.
If I start Xournal++ and start to draw, the pressure detection of my graphic tablet stops working in Xournal++. For the first line it works great, but if I stop drawing and draw another line, it has a fixed thickness and can't be changed by pressing more or less.

**To Reproduce**
1. Start Xournal++
2. Draw your first line and play around with pen pressure (should work)
3. Draw a second line and play around with pen pressure (should work too, but doesn't work)

**Expected behavior**
The pen pressure detection should work like in version 1.0.20 or in older development builds of 1.1.0 (I don't exactly know which of them worked). In this versions I can draw every line with pen pressure detection (with the new input system). 

**Screenshots of Problem**
I've tried to capture a video, but it doesn't worked well and I can't show the problem very well in a video.

**Additional context**
While trying to capture a video of this problem, I've found out that this problem temporarily disappears if I change the active window, controle it with my pen and then start to draw in Xournal++ while it's inactive (it gets activated instantly and everything works)... I've also tried to update my Wacom driver

**Comment by LittleHuba** [comment]
This is most likely related to the switch to the new version of GTK. Did you try to enable "Windows Ink" in the Wacom driver?

**Comment by Nick768** [comment]
Under Windows 7 there isn't an option to enable Windows Ink, because it's a Microsoft feature which doesn't exist in 7 as far as I know. Under Windows 11 I've tried it again now (without enabling Windows Ink, because I can't find it xD) and - I can't explain why - it works! The only thing that doesn't work are my pen buttons; maybe I'm too stupid to configure them correctly? But on my Windows 7 machine the problem still exists. Are there known incompatibilities between the newest GTK and Windows 7?

**Comment by LittleHuba** [comment]
@lb90 is the one that can answer these questions. He is the main driver behind the improved support for Windows.

**Comment by Nick768** [comment]
OK. Thanks for your help. I will wait if he answers my questions here. If not I will ask him directly. But to have a good workflow: is there a way to download older versions with the older GTK version? If I click on any tag in the releases section there is only the latest release available. And the 1.0.20 isn't useable for my purposes, because there are features from 1.1.0 missing I'm using quite often.

**Comment by lb90** [comment]
Hello, everyone! :)

Yes, GTK, since version 3.24.30, defaults to Windows Ink on all platforms where it is supported (Windows 8 and newer). As @Nick768 stated, Windows 7 does not have Windows Ink, so on Windows 7 GTK still uses Wintab.

Now, in the latest releases of GTK some improvements for Wintab were landed, too, so probably it's one of those improvements that causes this regression.

> there a way to download older versions with the older GTK version?

Yes, you can download GTK3 binaries from the [MSYS2 repo](https://repo.msys2.org/mingw/mingw64/) (MSYS2 is used for building the official Windows Xournal++ installer). I'd try with:

* [gtk3-3.24.29-2](https://repo.msys2.org/mingw/mingw64/mingw-w64-x86_64-gtk3-3.24.29-2-any.pkg.tar.zst)
* [gtk3-3.24.29-1](https://repo.msys2.org/mingw/mingw64/mingw-w64-x86_64-gtk3-3.24.29-1-any.pkg.tar.zst)
* [gtk3-3.24.27-1](https://repo.msys2.org/mingw/mingw64/mingw-w64-x86_64-gtk3-3.24.27-1-any.pkg.tar.zst)

To decompress Zstandard archives I use https://github.com/mcmilk/7-Zip-zstd. When you have extracted the archives, you can find `mingw64\bin\libgtk-3-0.dll` and `mingw64\bin\libgdk-3-0.dll`. Copy those two DLLs inside the `bin `folder of Xournal++.

**Comment by Nick768** [comment]
@lb90 it works! I've downloaded gtk3-3.24.29-2 and it works perfectly under Windows 7! Thank you very much! And I've already known and used 7zip-Zstd ;-). I think this issue is solved now. I'll close it later this day. Maybe somebody else wants to ask something before it's closed? And if you need somebody to test something with Windows 7 or Windows 11 you could contact me (I will rebuild Xournal++ myself with input logs enabled)!

**Comment by Nick768** [comment]
Now I've tried to use the older version of GTK on Windows 11. It works! Now I can press my pen buttons again! Again: Thanks! My problem is solved! Now I close this issue... Althoug I'm interested in the details of this issue... Should I leave this issue open or open another, more specific issue? The idea is to discuss the problem and perhaps somebody could show me (and anybody else who is interested) the lines of code which are related to this issue. Then I could help finding the issue and test some things by myself (maybe not instantly because im short in time)...

**Comment by lb90** [comment]
Yes, let's leave this issue open for now. If you are willing to test, I can send you some links to the DLLs built by the automated CI infrastructure so that we can track the cause of this issue!

Of course you can do it whenever you have the time, no hurry! ;)

**Comment by Nick768** [comment]
> Of course you can do it whenever you have the time, no hurry! ;)

That's perfect :+1: 

I don't exactly know what DLLs you mean, but this sounds good. I guess you mean GTK related DLLs, right?

**Comment by lb90** [comment]
Yep.

**Comment by Nick768** [comment]
I've noticed that the 1.1.0-hotfix.1 version (commit 9d1277d) uses the old GTK+ version. Therefore it should work out of the box for me. I thought I've tested this version too... However, after testing this version works for me on both Windows systems. Am I the only person who have problems with GTK+ 3.24.30? Or is it because I am using an outdated system (Windows 7) and a too modern system (Windows 11 beta)? Are there problems with Windows 10 too (especially with Wacom products, because under Windows 11 the only problems are the graphic tablet buttons; they don't work with this version of GTK+)?

**Comment by lb90** [comment]
> I've noticed that the 1.1.0-hotfix.1 version (commit 9d1277d) uses the old GTK+ version. Therefore it should work out of the box for me. I thought I've tested this version too... However, after testing this version works for me on both Windows systems.

1.1.0-hotfix.1 comes with GTK 3.24.30. Are you saying it's working correctly on both systems?

**Comment by lb90** [comment]
Specifically, what are the problems with the graphics tablet buttons?

**Comment by Nick768** [comment]
> 1.1.0-hotfix.1 comes with GTK 3.24.30. Are you saying that it's working correctly on both systems?

No. Only the version without hotfix is working correctly. Therefore my question was if it could be operating system related (maybe because Windows 7 doesn't have Windows Ink and Windows 11 is in a beta). The version with the hotfix (and the latest nightly) doesn't work correctly (both with GTK 3.24.30). On Windows 7 only the pen pressure doesn't work.

> what are the problems with the graphics tablet buttons?

The problems only appear on Windows 11 and only in Xournal++ with GTK 3.24.30. The problem is, I can't map actions to my pen buttons. Typically I do this in the "Stylus" settings of Xournal++ and set button 1 to eraser and button 2 to hand. Then I have to go to my settings of the Wacom software and set the stylus buttons to set by application (I don't know the exact name of this setting in english, but this describes it best). This works with every program I've tested (I can press those buttons and the selected action get triggered) but not in Xournal++ with GTK 3.24.30.
And here I would ask the same question: Does it work on other operating systems (like Windows 10)? It is curios that two different Windows versions behave that different with the exact same settings. And I can't test it on a Windows 10 (or 8/8.1) machine, because I don't have Windows 8/8.1/10...

**Comment by CoccodrillooXDS** [comment]
> Are there problems with Windows 10 too (especially with Wacom products, because under Windows 11 the only problems are the graphic tablet buttons; they don't work with this version of GTK+)?

I'm using the 1.1.0-hotfix.1 version (commit 9d1277d) on Windows 10 Pro.
When I first opened Xournal++ I thought that my Wacom Bamboo Pen wouldn't work anymore (it's also a very very old Wacom tablet) with pressure and pen buttons.
I tried enabling Windows Ink (you should be able to enable that from Wacom software, theoretically) and it works now! (in my case I searched "Wacom Preferences" and my preferences for my Wacom tablet)
![image](https://user-images.githubusercontent.com/25722501/129443200-96b4e579-6ac9-443f-98db-e8bfc866e3a6.png)
(the screenshot is in Italian)

I don't think if you also have this preinstalled with your Wacom driver (since mine is a very old version (~2014))

**Comment by CoccodrillooXDS** [comment]
I searched online your Wacom tablet and no, you shouldn't have it because your software is a lot recent than mine 😅😅

Anyway, from your Wacom software you should try to open this window (image), go to "Mapping" and enable "Use Windows Ink".

![image](https://user-images.githubusercontent.com/25722501/129443674-10ba82f5-d2eb-4c7a-aff1-3e113b602c2d.png)

If I'm not mistaken Windows Ink exists also on Windows 7 (way back also to Windows XP Tablet Edition), maybe just with another name.

**Comment by Nick768** [comment]
Oh sorry, I wasn't online for a few days... And yes, you're right. I've found this setting on Windows 7 too. But it was already enabled... Perhaps I should try to deactivate it and report back if it works...

**Comment by lb90** [comment]
Hello! Sorry for the delay, I've been away for some time ;)

@CoccodrillooXDS is right: Windows Ink was introduced in XP. Anyway, GTK currently supports it through the modern interface (API) introduced in Windows 8, so for now Windows 8 or later is required to use Windows Ink.

As for the tablet buttons, yes: right now support is pretty basic. Anyway, I plan to work on that!

**Comment by Nick768** [comment]
> Windows Ink was introduced in XP

That's interesting! I don't know that. 

> Anyway I plan to work on that!

I would be able to test your work. And if I would know which files are responsible for the tablet input system I could fix some problems (I guesss).

**Comment by marc1uk** [comment]
I realise this is a very old topic, but I've just found this issue using xournal++ as a pdf annotator running on an old (but still very functional) windows 7 tablet. Unfortunately [the links to the dlls](https://github.com/xournalpp/xournalpp/issues/3311#issuecomment-894242781) are no longer online, and the versions on the wayback machine seem truncated - i get errors extracting them. If anyone has a copy, could they upload for posterity? 🙏

**Comment by bhennion** [comment]
Did you try newer versions like https://repo.msys2.org/mingw/mingw64/mingw-w64-x86_64-gtk3-3.24.50-1-any.pkg.tar.zst ?

**Edit:** maybe you actually need old versions... If so, then you can download old Xournal++ installers from our release page - they include builds of gtk3.

**Comment by marc1uk** [comment]
I dunno why I didn't think of that! 🙈
<img width="379" height="186" alt="Image" src="https://github.com/user-attachments/assets/10d6fe3a-b354-4c91-8855-33221d938815" />
Thankyou. :D 

I actually found that:
* 1.1.0 pressure sensitivity works fine
* 1.1.0-hotfix (first version with libgtk-3.24) senses pressure only the first time the pen is touched to the tablet, but then the pen has to move entirely out of the field of the screen (>10cm away) before it works again for the next stroke. Curious, as it _can_ work... 
* current version (1.2.8) pressure sensitivity does not work at all.

I tried copying the libgdk-3.0.dll to the current version, and in that case pressure sensitivity works fully.... however, touching the File / Edit / ... menu at the top with the stylus immediately crashes the application. 🤔 
I could do further testing as it may be possible to get later versions working with pressure, but I'll leave that for a future date - for now v1.1.0 is sufficient.

**Comment by bhennion** [comment]
Maybe you can build your own Xournal++, using the old libs, but it will probably take some effort to setup the build environment.

**Comment by marc1uk** [comment]
possibly, current MSYS2 doesn't support win7 - they do mention the latest version that does, but whether it will work for building the latest xournal++ would require testing. For the time being this seems to have enough features for my use.
